indexer/etl/l1_etl.go

```diff
@@ -20,43 +19,49 @@ type L1ETL struct {
 	db *database.DB
 }
 
-func NewL1ETL(log log.Logger, db *database.DB, client node.EthClient, contracts config.L1Contracts) (*L1ETL, error) {
+// NewL1ETL creates a new L1ETL instance that will start indexing from different starting points
+// depending on the state of the database and the supplied start height.
+func NewL1ETL(log log.Logger, db *database.DB, client node.EthClient, start *big.Int,
+	contracts config.L1Contracts) (*L1ETL, error) {
 	log = log.New("etl", "l1")
 
-	contractValue := reflect.ValueOf(contracts)
-	fields := reflect.VisibleFields(reflect.TypeOf(contracts))
-	l1Contracts := make([]common.Address, len(fields))
-	for i, field := range fields {
-		// ruleid: unsafe-reflect-by-name
-		addr, ok := (contractValue.FieldByName(field.Name).Interface()).(common.Address)
-		if !ok {
-			log.Error("non-address found in L1Contracts", "name", field.Name)
-			return nil, errors.New("non-address found in L1Contracts")
-		}
-
-		log.Info("configured contract", "name", field.Name, "addr", addr)
-		l1Contracts[i] = addr
+	latestHeader, err := db.Blocks.L1LatestBlockHeader()
+	if err != nil {
+		return nil, err
 	}
 
-	latestHeader, err := db.Blocks.L1LatestBlockHeader()
+	cSlice, err := contracts.AsSlice()
 	if err != nil {
 		return nil, err
 	}
 
+	// Determine the starting height for traversal
 	var fromHeader *types.Header
 	if latestHeader != nil {
 		log.Info("detected last indexed block", "number", latestHeader.Number.Int, "hash", latestHeader.Hash)
 		fromHeader = latestHeader.RLPHeader.Header()
-	} else {
-		log.Info("no indexed state, starting from genesis")
+
+	} else if latestHeader != nil && latestHeader.Number.Int.Cmp(start) == 0 {
+		log.Info("no indexed state in storage, starting from L1 genesis")
+
+	} else { // latestHeader == nil
+		log.Info("no indexed state in storage, starting from supplied L1 height", "height", start.String())
+		header, err := client.BlockHeaderByNumber(start)
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch starting block header: %w", err)
+		}
+
+		fromHeader = header
 	}
```

Comment:
The first part of this `else if` condition isn't needed since we know that `latestHeader == nil` after the first `if` check. How about this change:


```suggestion
	} else if start.Bitlen() > 0 {
	log.Info("no indexed state in storage, starting from supplied L1 height", "height", start.String())
		header, err := client.BlockHeaderByNumber(start)
		if err != nil {
			return nil, fmt.Errorf("could not fetch starting block header: %w", err)
		}

		fromHeader = header
	} else {
		log.Info("no indexed state in storage or starting height, starting from L1 genesis")
    }
```

Having the last case be the simplest case (genesis) reads a bit better imo

Reply:
[dd4f561](https://github.com/ethereum-optimism/optimism/pull/6914/commits/dd4f561552f16d8de1e09cc46f39079d2bdcb77d)